rust/arrow/src/array/equal/primitive.rs

```diff
@@ -32,7 +36,10 @@ pub(super) fn primitive_equal<T>(
     let lhs_values = &lhs.buffers()[0].as_slice()[lhs.offset() * byte_width..];
     let rhs_values = &rhs.buffers()[0].as_slice()[rhs.offset() * byte_width..];
 
-    if lhs.null_count() == 0 && rhs.null_count() == 0 {
+    let lhs_null_count = count_nulls(lhs_nulls, lhs_start, len);
+    let rhs_null_count = count_nulls(rhs_nulls, rhs_start, len);
+
+    if lhs_null_count == 0 && rhs_null_count == 0 {
```

Comment:
This is even better: if the buffer only has 1s outside of the range, we short-circuit to the fast case. 👍